Feature: 21 orchestrator rpc client for blocks confirmation flow

mxmar opened the pull request:
**What type of PR is this?**

Feature

**What does this PR do? Why is it needed?**

This feature provides `orchestrator` RPC client in `vanguard` beacon-chain client. It's required to achieve `orc_confirmVanBlockHashes` from https://gist.github.com/frozeman/d52090f3b8c4b654cacc279d894df31c document.

**Which issues(s) does this PR fix?**

Fixes https://github.com/lukso-network/vanguard-consensus-engine/issues/21

**Other notes for review**

mxmar committed:
Feature: 21 orchestrator rpc client for blocks confirmation flow
- added client package implementation;
- missing unit tests and orc rpc client invoke (with Dial);

Signed-off-by: mxmar <macht4tech@gmail.com>

atif-konasl commented on beacon-chain/orchestrator/client.go:
```diff
@@ -0,0 +1,57 @@
+	orcRpc.client.Close()
+}
+
+func (orcRpc *RPCClient) ConfirmVanBlockHashes(ctx context.Context, blockHashes []*BlockHash) ([]*BlockStatus, error) {
```
@mxmar Method definition should be `ConfirmVanBlockHashes(ctx context.Context, request []*vanTypes.ConfirmationReqData) (response []*vanTypes.ConfirmationResData, err error)` and in this method, you just convert common.Hash to [32] byte for hash

mxmar replied:
Ok, so conversion logic will lie in this method as you has suggested.

atif-konasl replied:
yes

mxmar replied:
done

atif-konasl commented on beacon-chain/orchestrator/client.go:
```diff
@@ -0,0 +1,57 @@
+
+const confirmVanBlockHashesMethod = "orc_confirmVanBlockHashes"
+
+type Client interface {
```
@mxmar Please define another interface for method `ConfirmVanBlockHashes(ctx context.Context, request []*vanTypes.ConfirmationReqData) (response []*vanTypes.ConfirmationResData, err error)` and embed the interface into this Client interface.

mxmar replied:
@atif-konasl In other words you would like to separate RPC logic and RPC requests like `rpc please confirm these blocks` right?

atif-konasl replied:
yes right. `blockchain` package will not invoke these `Dial`, `DialContext` etc methods. So you just need an interface which only have this `ConfirmVanBlockHashes(ctx context.Context, request []*vanTypes.ConfirmationReqData) (response []*vanTypes.ConfirmationResData, err error)` method definition so that `blockchain` package can call this method.

atif-konasl replied:
Okay

mxmar replied:
got it, done.

atif-konasl commented on beacon-chain/orchestrator/client.go:
```diff
@@ -0,0 +1,57 @@
+	DialContext(context.Context, string) (*Client, error)
+	NewClient(*rpc.Client) *Client
+	Close()
+	ConfirmVanBlockHashes(context.Context, []*BlockHash) ([]*BlockStatus, error)
```
@mxmar Method definition should be `ConfirmVanBlockHashes(ctx context.Context, request []*vanTypes.ConfirmationReqData) (response []*vanTypes.ConfirmationResData, err error)` and in this method, you just convert common.Hash to [32] byte for hash

mxmar replied:
Ok, so conversion logic will lie in this method as you has suggested. I will prepare separate interface for it.

mxmar replied:
done.

blazejkrzak commented on beacon-chain/orchestrator/client_test.go:
```diff
@@ -0,0 +1,7 @@
+import "testing"
+
+func TestRPCClient_ConfirmVanBlockHashes(t *testing.T) {
+
```
wiggle

mxmar replied:
tests have been refactored to proper one

mxmar committed:
Feature: 21 orchestrator rpc client for blocks confirmation flow
- modified BUILD bazel file;
- added connectivity test;
- missing unit tests and orc rpc client invoke (with Dial);

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 21 orchestrator rpc client for blocks confirmation flow
- modified BUILD bazel file;
- added vanTypes conversion;
- added initial mock rpc server;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
spike

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 21 orchestrator rpc client for blocks confirmation flow
- modified go test for rcpClient;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 21 orchestrator rpc client for blocks confirmation flow
- modified github action yml;

Signed-off-by: mxmar <macht4tech@gmail.com>

mxmar committed:
Feature: 21 orchestrator rpc client for blocks confirmation flow
- modified go test for rcpClient;

Signed-off-by: mxmar <macht4tech@gmail.com>
